db/hints/internal/hint_endpoint_manager.cc

```diff
@@ -53,46 +53,47 @@ bool hint_endpoint_manager::store_hint(schema_ptr s, lw_shared_ptr<const frozen_
             // We discard the future to this call to `with_gate()`, so wrapping this lambda
             // within `coroutine::lambda` will NOT work. That's why we need to copy these
             // variables by hand.
+            hint_endpoint_manager& self = *this;
```

Comment:
so we are suffering from https://github.com/scylladb/seastar/blob/master/doc/lambda-coroutine-fiasco.md#alternative-solution-when-lifetime-extension-cannot-be-used here.

have you considered naming the lambda passed to `with_gate()`, and reference it using `std::ref()`, like:
```c++
auto do_store_hint = [this, s = std::move(s), fm = std::move(fm), tr_state = tr_state] () mutable -> future<> {
    // ....
    // could yield
    // references the captured variables
};
try {
    (void) with_gate(_store_gate, std::ref(do_store_hint));
} catch (...) {
    // 
    return false;
}
return true;
```

IMHO, this approach is better in the sense that we don't have to use `self` where `this` is typically used, it's confusing without context.

Reply:
Exactly! The suspicious thing is ASAN doesn't detect it, so either I misunderstand the semantics here, or we're suffering from UB (seems very unlikely). No matter what it is, I think it's better to get that pointer and be safe (the compiler should optimise it anyway).